File: WU/internal/ui/wrangler.go
package ui

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"golang.org/x/term"
)

// Colors
var (
	gray   = color.New(color.FgHiBlack).SprintFunc()
	cyan   = color.New(color.FgCyan).SprintFunc()
	bold   = color.New(color.Bold).SprintFunc()
	blue   = color.New(color.FgBlue).SprintFunc()
	green  = color.New(color.FgGreen).SprintFunc()
	yellow = color.New(color.FgYellow).SprintFunc()
	orange = color.RGB(189, 91, 8).SprintFunc() // #bd5b08
	red    = color.New(color.FgRed).SprintFunc()
)

const line = "───────────────────"

type StepCtx struct {
	Title   string
	Current int
	Total   int
}

// Banner prints the tool name and version in Wrangler style
func Banner(tool, version string) {
	fmt.Printf("\n %s %s %s\n", cyan("⛅️"), bold(tool), gray(version))
	fmt.Println(gray(line))
}

// Section prints a section header
func Section(step StepCtx) {
	fmt.Println("")
	fmt.Printf("%s %s %s\n", gray("╭"), orange(step.Title), gray(fmt.Sprintf("Step %d of %d", step.Current, step.Total)))
	fmt.Println(gray("│"))
}

// Item prints an item in the section
// If value is empty, it just prints the label
// If value is provided, it prints label then indented value
func Item(label string, value ...string) {
	fmt.Printf("%s %s\n", gray("├"), label)
	for _, v := range value {
		fmt.Printf("%s %s %s\n", gray("│"), gray("dir"), v) // "dir" is hardcoded in example but maybe we want generic?
		// keeping generic:
		// fmt.Printf("%s %s\n", gray("│"), v)
	}
	fmt.Println(gray("│"))
}

// ItemValue prints a key-value pair
func ItemValue(key, value string) {
	fmt.Printf("%s %s\n", gray("├"), key)
	fmt.Printf("%s %s %s\n", gray("│"), gray("value"), value)
	fmt.Println(gray("│"))
}

// EndLine prints the closing line of a section
func EndLine(label string) {
	fmt.Printf("%s %s\n", gray("╰"), label)
}

func Info(msg string) {
	fmt.Printf("%s %s\n", blue("ℹ"), msg)
}

func Ok(msg string) {
	fmt.Printf("%s %s\n", green("✅"), msg)
}

func Warn(msg string) {
	fmt.Printf("%s %s\n", yellow("⚠️"), msg)
}

func Fail(msg string) {
	fmt.Printf("%s %s\n", red("❌"), msg)
}

// Spinner runs a task with a spinner
func Spin(label string, task func() error) error {
	s := spinner.New(spinner.CharSets[14], 100*time.Millisecond) // Dots
	s.Suffix = " " + label
	s.Color("cyan")
	s.Start()
	err := task()
	s.Stop()
	if err != nil {
		return err
	}
	return nil
}

// Prompt prompts the user for input with the Wrangler style
// Example:
// ├ In which directory do you want to create your application?
// │ dir ./my-worker
func Prompt(question string, def string) string {
	fmt.Printf("%s %s\n", gray("├"), question)

	prefix := "value"
	// To match wrangler "dir ./my-worker" style where user types after "dir "
	// We'll just use a generic prompt.

	promptLine := fmt.Sprintf("%s %s ", gray("│"), gray(prefix))
	if def != "" {
		promptLine = fmt.Sprintf("%s %s [%s] ", gray("│"), gray(prefix), def)
	}

	fmt.Print(promptLine)

	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	input = strings.TrimSpace(input)

	if input == "" {
		return def
	}

	// Print the vertical line to close the look if we want, but usually prompt ends the interaction for that line.
	// But in Wrangler example, it continues.
	fmt.Println(gray("│"))

	return input
}

func PromptSecret(question string) string {
	fmt.Printf("%s %s\n", gray("├"), question)
	fmt.Printf("%s %s ", gray("│"), gray("secret"))

	bytePassword, err := term.ReadPassword(int(os.Stdin.Fd()))
	fmt.Println() // Newline after password input
	fmt.Println(gray("│"))

	if err != nil {
		return ""
	}
	return string(bytePassword)
}

func PromptYesNo(question string, def bool) bool {
	fmt.Printf("%s %s\n", gray("├"), question)

	yv := "y/N"
	if def {
		yv = "Y/n"
	}

	fmt.Printf("%s %s (%s) ", gray("│"), gray("confirm"), yv)

	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	input = strings.ToLower(strings.TrimSpace(input))

	fmt.Println(gray("│"))

	if input == "" {
		return def
	}
	return input == "y" || input == "yes"
}

File: WU-Nodejs/src/ui/wrangler.ts
import chalk from "chalk";
import inquirer from "inquirer";

export type StepCtx = { title: string; current: number; total: number };

const line = "───────────────────";

export function banner(tool: string, version: string) {
  // 模仿 wrangler：云朵 + 版本 + 分割线
  console.log(`\n ${chalk.cyan("⛅️")} ${chalk.bold(tool)} ${chalk.gray(version)}`);
  console.log(chalk.gray(line));
}

export function section(step: StepCtx) {
  // “╭ … Step x of y”
  console.log("");
  console.log(
    `${chalk.gray("╭")} ${chalk.bold(step.title)} ${chalk.gray(`Step ${step.current} of ${step.total}`)}`
  );
  console.log(chalk.gray("│"));
}

export function item(label: string, value?: string) {
  // “├ label / │ value …”
  if (value === undefined) {
    console.log(`${chalk.gray("├")} ${label}`);
  } else {
    console.log(`${chalk.gray("├")} ${label}`);
    console.log(`${chalk.gray("│")} ${chalk.gray("dir")} ${value}`);
  }
  console.log(chalk.gray("│"));
}

export function endLine(label: string) {
  // “╰ …”
  console.log(`${chalk.gray("╰")} ${label}`);
}

export function info(msg: string) {
  console.log(`${chalk.blue("ℹ")} ${msg}`);
}

export function ok(msg: string) {
  console.log(`${chalk.green("✅")} ${msg}`);
}

export function warn(msg: string) {
  console.log(`${chalk.yellow("⚠️")} ${msg}`);
}

export function fail(msg: string) {
  console.error(`${chalk.red("❌")} ${msg}`);
}

export async function promptText(message: string, def?: string) {
  const ans = await inquirer.prompt([
    { type: "input", name: "v", message, default: def ?? undefined }
  ]);
  return String(ans.v ?? "").trim();
}

export async function promptPassword(message: string) {
  const ans = await inquirer.prompt([
    { type: "password", name: "v", message, mask: "*" }
  ]);
  return String(ans.v ?? "");
}

export async function promptConfirm(message: string, def = true) {
  const ans = await inquirer.prompt([
    { type: "confirm", name: "v", message, default: def }
  ]);
  return !!ans.v;
}
